### PR TITLE
Fix extract local function crash

### LINKED ISSUE #45422: Editor: crashed in ExtractMethod
_This issue has been moved from [a ticket on Developer Community](https://developercommunity.visualstudio.com/content/problem/1087377/editor-crashed-in-extractmethod.html)._

---
C#

Editing my file by copy/pasting code from another window. 

For sure, I was not actually extracting a method (just copying lines).

Got this crash:

```
System.InvalidOperationException : Sequence contains no elements
   at System.Linq.Enumerable.Single[TSource](IEnumerable`1 source)
   at Microsoft.CodeAnalysis.ExtractMethod.InsertionPoint.ComputeContextNode()
   at System.Lazy`1.CreateValue()
   at System.Lazy`1.LazyInitValue()
   at System.Lazy`1.get_Value()
   at Microsoft.CodeAnalysis.ExtractMethod.InsertionPoint.GetContext()
   at async Microsoft.CodeAnalysis.ExtractMethod.MethodExtractor.CodeGenerator`3.GenerateAsync[TStatement,TExpression,TNodeUnderContainer](<Unknown Parameters>)
   at async Microsoft.CodeAnalysis.ExtractMethod.MethodExtractor.ExtractMethodAsync(<Unknown Parameters>)
   at async Microsoft.CodeAnalysis.ExtractMethod.AbstractExtractMethodService`3.ExtractMethodAsync[TValidator,TExtractor,TResult](<Unknown Parameters>)
   at async Microsoft.CodeAnalysis.CodeRefactorings.ExtractMethod.ExtractMethodCodeRefactoringProvider.ExtractLocalFunctionAsync(<Unknown Parameters>)
   at async Microsoft.CodeAnalysis.CodeRefactorings.ExtractMethod.ExtractMethodCodeRefactoringProvider.GetCodeActionsAsync(<Unknown Parameters>)
   at async Microsoft.CodeAnalysis.CodeRefactorings.ExtractMethod.ExtractMethodCodeRefactoringProvider.ComputeRefactoringsAsync(<Unknown Parameters>)
   at async Microsoft.CodeAnalysis.CodeRefactorings.CodeRefactoringService.GetRefactoringFromProviderAsync(<Unknown Parameters>)
   at Microsoft.VisualStudio.Telemetry.WindowsErrorReporting.WatsonReport.GetClrWatsonExceptionInfo(Exception exceptionObject)
```

---
#### Visual Studio Feedback System on 6/22/2020, 07:26 PM: 

<p>We have directed your feedback to the appropriate engineering team for further evaluation. The team will review the feedback and notify you about the next steps.</p>

---
### Original Solutions
(no solutions)

### PR DESCRIPTION
**Pull request by allisonchou**
Fixes #45422.

Extract local function always introduces a call to the newly extracted function. There are some cases when introducing this call is impossible (e.g. when the selection is solely a local function), and trying to extract these selections currently results in a crash. We should not offer extract local function in this case.

### CODE REVIEW

**Comment by allisonchou** [comment]
Whoops hold on, noticed something weird with the fix - will re-open when ready.

File: src/EditorFeatures/DiagnosticsTestUtilities/CodeActions/AbstractCodeActionOrUserDiagnosticTest.cs [hunk 1]
```diff
@@ -255,21 +255,31 @@ private TestParameters WithScriptOptions(TestParameters parameters)
...
                 var offeredActions = Environment.NewLine + string.Join(Environment.NewLine, actions.Select(action => action.Title));
-                Assert.True(actions.Length == 0, "An action was offered when none was expected. Offered actions:" + offeredActions);
+
+                if (codeActionIndex == 0)
+                {
+                    Assert.True(actions.Length == 0, "An action was offered when none was expected. Offered actions:" + offeredActions);
+                }
+                else
+                {
+                    Assert.True(actions.Length <= codeActionIndex, "An action was offered at the specified index when none was expected. Offered actions:" + offeredActions);
```
Comment (davidwengier) [inline src/EditorFeatures/DiagnosticsTestUtilities/CodeActions/AbstractCodeActionOrUserDiagnosticTest.cs:281]: The message is the wrong way around, isn't it?

```suggestion
                    Assert.True(actions.Length <= codeActionIndex, "An action was expected at the specified index when none was offered. Offered actions:" + offeredActions);
```

File: src/Features/Core/Portable/ExtractMethod/InsertionPoint.cs [hunk 2]
```diff
@@ -2,7 +2,7 @@
 // The .NET Foundation licenses this file to you under the MIT license.
 // See the LICENSE file in the project root for more information.
 
-#nullable disable
+#nullable enable
```
Comment (davidwengier) [inline src/Features/Core/Portable/ExtractMethod/InsertionPoint.cs:5]: The whole line can be removed, and it will be equivalent.

**Review by CyrusNajmabadi — APPROVED** [review]

**Review by davidwengier — APPROVED** [review]
A general question: Does this still generate the local function, and just show the message as information, or is this more of an error that says "We can't do anything: <reason>" ?

File: src/EditorFeatures/DiagnosticsTestUtilities/CodeActions/AbstractCodeActionOrUserDiagnosticTest.cs (hunk 1 above)
Comment (allisonchou) [reply src/EditorFeatures/DiagnosticsTestUtilities/CodeActions/AbstractCodeActionOrUserDiagnosticTest.cs:281]: Hmm, I think the existing way is correct? The method is testing to make sure that there's no code action available at the specified index.

To address your other question - this fix makes it so we don't offer the extract local function in this scenario.

Comment (davidwengier) [reply src/EditorFeatures/DiagnosticsTestUtilities/CodeActions/AbstractCodeActionOrUserDiagnosticTest.cs:281]: Oh whoops, you're absolutely right!

**Pull request merged**